src/Workspaces/Core/Portable/Workspace/Solution/SourceInfoProvider.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface ISourceFileInfoProvider
+    {
+        /// <summary>
+        /// return SourceFileInfo for the given full path, if this provide doesn't handle the given file
+        /// return null.
+        /// </summary>
+        Task<SourceFileInfo> GetSourceFileInfoAsync(ProjectId projectId, string projectFullPath, string fileFullPath, CancellationToken cancellationToken);
```

Comment:
@jasonmalinowski - I asked for this. Since this service is MEF-imported globally, we need to know what project you're asking for. We create the `SourceFileInfo` when you ask for it, and then we have to associated it later with our state.

Reply:
let me know if you guys want to change this.